google/cloud/internal/rest_retry_loop.h

```diff
@@ -61,41 +62,35 @@ template <
     typename std::enable_if<google::cloud::internal::is_invocable<
                                 Functor, RestContext&, Request const&>::value,
                             int>::type = 0>
-auto RestRetryLoopImpl(std::unique_ptr<RetryPolicy> retry_policy,
-                       std::unique_ptr<BackoffPolicy> backoff_policy,
+auto RestRetryLoopImpl(RetryPolicy& retry_policy, BackoffPolicy& backoff_policy,
                        Idempotency idempotency, Functor&& functor,
                        Request const& request, char const* location,
                        Sleeper sleeper)
     -> google::cloud::internal::invoke_result_t<Functor, RestContext&,
                                                 Request const&> {
-  Status last_status;
-  while (!retry_policy->IsExhausted()) {
+  auto last_status = internal::DeadlineExceededError(
+      "Retry policy exhausted before first request attempt", GCP_ERROR_INFO());
+  while (!retry_policy.IsExhausted()) {
     RestContext rest_context;
     auto result = functor(rest_context, request);
-    if (result.ok()) {
-      return result;
-    }
+    if (result.ok()) return result;
     last_status = internal::GetResultStatus(std::move(result));
     if (idempotency == Idempotency::kNonIdempotent) {
       return internal::RetryLoopError("Error in non-idempotent operation",
                                       location, last_status);
     }
-    if (!retry_policy->OnFailure(last_status)) {
-      // The retry policy is exhausted or the error is not retryable, either
-      // way, exit the loop.
-      break;
-    }
-    sleeper(backoff_policy->OnCompletion());
-  }
-  if (!retry_policy->IsExhausted()) {
-    // The last error cannot be retried, but it is not because the retry
-    // policy is exhausted, we call these "permanent errors", and they
-    // get a special message.
-    return internal::RetryLoopError("Permanent error in", location,
-                                    last_status);
+    // The retry policy is exhausted or the error is not retryable. Either
+    // way, exit the loop.
+    if (!retry_policy.OnFailure(last_status)) break;
+    sleeper(backoff_policy.OnCompletion());
   }
-  return internal::RetryLoopError("Retry policy exhausted in", location,
-                                  last_status);
+  // The last error cannot be retried, but it is not because the retry
+  // policy is exhausted. We call these "permanent errors", and they
+  // get a special message.
+  char const* prefix = !retry_policy.IsExhausted()
+                           ? "Permanent error in"
+                           : "Retry policy exhausted in";
+  return internal::RetryLoopError(prefix, location, last_status);
```

Comment:
observation: in the (rare) exhausted before start case, the error message is a bit redundant:

```
Retry policy exhausted in <location>:  Retry policy exhausted before first request attempt
```

I am fine with this.

Reply:
I think I will open a separate thread to ~shave this baby yak~ cleanup the error messages from retry loops.